### DISCUSSION TITLE
consider not linking Python by default

### DISCUSSION THREAD

**Discussion by glyph** [Homebrew/discussions#5517] [Tap maintenance and Homebrew development] [answered]
### Output of `brew config`

```shell
N/A
```

### Output of `brew doctor`

```shell
N/A
```

### Description of issue

As set out in https://justinmayer.com/posts/homebrew-python-is-not-for-you/ , homebrew python is a support structure for homebrew formulae, not really configured in a way which makes it suitable for use as a user-facing development environment.  In particular its upgrade cadence makes it a frequent source of complaints when users' virtual environments break.

I think that it would be a benefit to make Homebrew's python not get linked by default, so that this intention is a bit more obvious, and users do not come to inadvertently depend on a non-reproducible installation of Python from Homebrew that they have installed only by accident as a dependency of another tool.

**Comment by SMillerDev** [comment]
That's not an article from Homebrew though. I'd suggest that Homebrew Python is akin to Arch Linux python. You can't depend on it always being the same version, but if you need a python version it'll do just fine.

**Reply by glyph** [reply]
> That's not an article from Homebrew though.

I didn't mean to say it was first-party advice, only that it was very common advice that I've heard repeated very frequently.

> I'd suggest that Homebrew Python is akin to Arch Linux python. You can't depend on it always being the same version, but if you need a python version it'll do just fine.

If python.org shipped first-party builds of Python for Linux distributions, I'd also recommend that they keep their own Python off the path.  But they don't, so it's up to the distro. In the case of macOS though, there are high-quality first-party builds.  (An alternate solution here is to make Homebrew's Python be a cask, which would be even better, but I have to assume there are some good reasons Homebrew builds its own.)

**Reply by p-linnane** [reply]
There are multiple reasons we build our own Python and a lot of work goes into maintaining it. Only supporting Python as a cask would mean we cannot support anything depending on Python in Homebrew/core. Formulae in Homebrew/core must be built from source, and in turn the formulae they depend on must be built from source. Formulae cannot depend on casks, because it violates this fundamental tenant of Homebrew.

We'll ship first party binaries as casks when it's not prudent or possible to build from source. We also have [a small blurb](https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap#upstream-taps) about our stance on upstreams trying to 'maintain control' of open source software distribution.

**Reply by glyph** [reply]
> Only supporting Python as a cask would mean we cannot support anything depending on Python in Homebrew/core

Yep, I figured that it was something like this. I want to be clear that that is *not* what I am asking for here.  As I said:

> > I have to assume there are some good reasons Homebrew builds its own.

I just want Homebrew's Python to be more clearly telegraphed as supporting Homebrew formulae, and not a good default for users doing their own Python development.  For those who *do* want Homebrew's python, it's just a `brew link python` away.

**Reply by SMillerDev** [reply]
> I just want Homebrew's Python to be more clearly telegraphed as supporting Homebrew formulae, and not a good default for users doing their own Python development.

I disagree here, it's been good for any Python development I've done where the newest Python version was beneficial.

If my colleagues were to ask for the best way to get a stable development environment I'd recommend Docker or a language version manager though. Just like with every other programming language. That doesn't mean that brew shouldn't link any programming languages.

**Reply by glyph** [reply]
In addition to the piece I linked already, here are some other opinions about Python and Homebrew:

- from the mac install guide, it is "complicated and trouble-prone" for more than simple scripts https://mac.install.guide/python/#:~:text=complicated%20and%20trouble-prone
- from some random Reddit commenters a few years ago, "do not use homebrew for this for all the reasons you have listed" https://www.reddit.com/r/learnpython/comments/ps1byc/comment/hdn9wmz/
- I have written my own piece on this; mentioning it because while it doesn't have much new (it mostly just cites Mayer) it has been pretty popular https://blog.glyph.im/2023/08/get-your-mac-python-from-python-dot-org.html
- "yet another issue with a homebrew update" https://apple.stackexchange.com/questions/470627/yet-another-issue-with-a-homebrew-update-python-upgrade-no-module-named
- the first party documentation says "Important: Python may be upgraded to a newer version at any time. Consider using a version manager such as pyenv if you require stability" https://docs.brew.sh/Homebrew-and-Python#python-3
- "do not use Homebrew Python for development" https://aviaryan.com/blog/python-mac-setup

I could go on. This is a common pain point for Python developers, who would be better served by a single unified set of instructions, "just get it from python.org", rather than thousands of people rediscovering that Homebrew is not appropriate for Python development one at a time because it doesn't support shareable version locking or synchronized-upgrading schemes.

**Reply by glyph** [reply]
Again, we're just talking about a default. I'm not saying that Homebrew is never the right choice, but it is clearly the *wrong* choice often enough that it seems like requiring the extra explicit step of `brew link` would give us the best of both worlds.

**Reply by SMillerDev** [reply]
I think we have different views of the default. 

I see the default as someone who is new to programming and just wants a python version because they're reading a book or something about python. They do `brew install python` and they have a working python version that they can play with.

It seems you see the default as a professional python developer who needs to be forced to use pyenv rather than the homebrew python.

---

I still believe my original answer stands, since it's really no different from the situation on Linux:  

> I'd suggest that Homebrew Python is akin to Arch Linux python. You can't depend on it always being the same version, but if you need a python version it'll do just fine.

**Reply by SMillerDev** [reply, marked as answer]
An additional note: you can always make a pull request to see what the maintainers of Homebrew (other than the two that responded here) think.

**Reply by glyph** [reply]
> An additional note: you can always make a pull request to see what the maintainers of Homebrew (other than the two that responded here) think.

OK. I thought that a discussion might be more appropriate, but I'll put up a PR when I have a chance.

**Reply by quackerex** [reply]
> You can't depend on it always being the same version, but if you need a python version it'll do just fine.

Unless you need very specific version of python, homebrew works fine imo. you can install specific version like `python@3.x`. You can install multiple versions and use `python3.x` in terminal to use the version you want(and also auto create and source venvs with `direnv`).


I agree with @glyph that it is weird `python` is not keg-only. The same way `ruby` is keg-only to avoid messing with system ruby, it make sense `python` to be keg-only

**Reply by SMillerDev** [reply]
> The same way ruby is keg-only to avoid messing with system ruby, it make sense python to be keg-only

There is no system python, same way there is no system Java. It's just a wrapper that points to something if it's installed (for python it points to the Command Line Tools)

**Reply by glyph** [reply]
> for python it points to the Command Line Tools

This may not be a relevant detail, but it points to the selected Xcode, which can be the Xcode CLT, but if you have Xcode proper installed it will be that, following the usual rules

```console
$ xcode-select -p
/Applications/Xcode.app/Contents/Developer
$ /usr/bin/python3 -c 'import sys; print(sys.executable)'
/Applications/Xcode.app/Contents/Developer/usr/bin/python3
```

This is only *possibly* relevant because if you have multiple Xcodes installed, it's possible that you can switch what Python version you're getting from `/usr/bin/python3` by changing your active Xcode.

**Reply by SMillerDev** [reply]
Ah, I thought it was hardcoded to CLT. Makes sense that it follows the system direction of where to look for developer tools.

**Reply by glyph** [reply]
> Ah, I thought it was hardcoded to CLT.

Glad I can help correct some misconceptions then :).

> Makes sense that it follows the system direction of where to look for developer tools.

Yeah, from looking at `ls -i`, it seems that cc, python3, lldb et. al. are all just hard links to the same shim.
